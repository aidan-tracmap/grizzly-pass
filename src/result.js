class Result {
  constructor(type, value) {
    this.type = type;
    this.value = value;
  }

  isOk() {
    return this.type === "ok";
  }

  isErr() {
    return this.type === "error";
  }

  mapErr(f) {
    return this.isErr() ? Err(f(this.value)) : this;
  }

  map(f) {
    return this.isOk() ? Ok(f(this.value)) : this;
  }

  andThen(f) {
    return this.isOk() ? f(this.value) : this;
  }

  orElse(f) {
    return this.isErr() ? f(this.value) : this;
  }

  unsafeUnwrap() {
    if (this.isOk()) return this.value;
    else throw new Error(`Unwrapped result with err: ${this.value}`);
  }
}

export function Err(value) {
  return new Result("error", value);
}

export function Ok(value) {
  return new Result("ok", value);
}

export function fromList(listOfResults) {
  const errs = listOfResults.filter(r => r.isErr());
  if (errs.length) return Err(errs.map(e => e.value));
  else return Ok(listOfResults.map(o => o.value));
}
