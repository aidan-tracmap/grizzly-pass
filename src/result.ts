export interface Result<T, E> {
  isOk(): boolean;
  isErr(): boolean;
  mapErr<E2>(f: (e: E) => E2): Result<T, E2>;
  map<U>(f: (t: T) => U): Result<U, E>;
  andThen<U, E2>(f: (t: T) => Result<U, E2>): Result<U, E | E2>;
  orElse<U, E2>(f: (e: E) => Result<U, E2>): Result<T | U, E2>;
  unsafeUnwrap(): T;
  unsafeUnwrapError(): E;
}

type InnerRep<T, E>
  = { type: "ok"; value: T }
  | { type: "error"; value: E };

class ResultImpl<T, E> implements Result<T, E> {

  inner: InnerRep<T, E>;

  constructor(inner_: InnerRep<T, E>) {
    this.inner = inner_;
  }

  isOk() {
    return this.inner.type === "ok";
  }

  isErr() {
    return this.inner.type === "error";
  }

  mapErr<E2>(f: (e: E) => E2) {
    switch (this.inner.type) {
      case "ok": return this;
      case "error": return Err(f(this.inner.value));
    }
  }

  map<U>(f: (t: T) => U) {
    switch (this.inner.type) {
      case "ok": return Ok(f(this.inner.value));
      case "error": return this;
    }
  }

  andThen<U>(f: (t: T) => Result<U, E>) {
    switch (this.inner.type) {
      case "ok": return f(this.inner.value);
      case "error": return this;
    }
  }

  orElse<E2>(f: (e: E) => Result<T, E2>) {
    switch (this.inner.type) {
      case "ok": return this;
      case "error": return f(this.inner.value);
    }
  }

  unsafeUnwrap() {
    switch (this.inner.type) {
      case "ok":
        return this.inner.value;
      case "error":
        throw new Error(`Unwrapped result with err: ${this.inner.value}`);
    }
  }

  unsafeUnwrapError() {
    switch (this.inner.type) {
      case "ok":
        throw new Error(`Unwrapped error of result with value: ${this.inner.value}`);
      case "error":
        return this.inner.value;
    }
  }
}

export function Err<E>(value: E): Result<any, E> {
  return new ResultImpl({ type: "error", value });
}

export function Ok<T>(value: T): Result<T, any> {
  return new ResultImpl({ type: "ok", value });
}

export function fromList<T, E>(listOfResults: Result<T, E>[]): Result<T[], E[]> {
  const errs = listOfResults.filter(r => r.isErr());
  if (errs.length) return Err(errs.map(e => e.unsafeUnwrapError()));
  else return Ok(listOfResults.map(o => o.unsafeUnwrap()));
}
