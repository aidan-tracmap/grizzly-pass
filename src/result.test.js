import { Err, Ok, fromList } from "./result";

it("returns the value if an ok result is unsafeUnwrap'd", () => {
  expect(Ok(113).unsafeUnwrap()).toBe(113);
});

it("throws if an error result is unsafeUnwrap'd", () => {
  expect(() => Err("bad").unsafeUnwrap()).toThrowErrorMatchingSnapshot();
});

it("leaves an ok result unchanged when mapErr is called", () => {
  expect(Ok(113).mapErr(i => i + 1)).toEqual(Ok(113));
});

it("changes an err result when mapErr is called", () => {
  expect(Err(113).mapErr(i => i + 1)).toEqual(Err(114));
});

it("changes an ok result when map is called", () => {
  expect(Ok(113).map(i => i + 1)).toEqual(Ok(114));
});

it("leaves an err result unchanged when map is called", () => {
  expect(Err(113).map(i => i + 1)).toEqual(Err(113));
});

it("users the returned result when flatMap is called", () => {
  expect(Ok(113).flatMap(i => Err("to error"))).toEqual(Err("to error"));
});

it("leaves an err result unchanged when flatMmap is called", () => {
  expect(Err(113).flatMap(i => Err(i + 1))).toEqual(Err(113));
});

describe("fromList", () => {
  it("returns an ok result if all list elements are ok results", () => {
    expect(fromList([Ok(1), Ok(2), Ok(3)])).toEqual(Ok([1, 2, 3]));
  });

  it("results an err result if all list elements are err results", () => {
    expect(fromList([Err(1), Err(2)])).toEqual(Err([1, 2]));
  });

  it("returns an err result if any list element is an err result", () => {
    expect(fromList([Ok(1), Err(2)])).toEqual(Err([2]));
  });
});
