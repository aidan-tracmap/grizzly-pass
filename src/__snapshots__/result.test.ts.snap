exports[`test throws if an error result is unsafeUnwrap'd 1`] = `"Unwrapped result with err: bad"`;
